drivers/include/net/netdev2.h

```diff
@@ -0,0 +1,180 @@
+ * 5. netdev2->event_callback() uses netdev2->driver.recv() to fetch packet
+ *
+ * @file
+ * @brief       Definitions low-level network driver interface
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @author      Martine Lenders <mlenders@inf.fu-berlin.de>
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ */
+
+#ifndef NETDEV2_H
+#define NETDEV2_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include <sys/uio.h>
+
+#include "net/netopt.h"
+
+enum {
+    NETDEV2_TYPE_UNKNOWN,
+    NETDEV2_TYPE_RAW,
+    NETDEV2_TYPE_ETHERNET,
+    NETDEV2_TYPE_802154,
+    NETDEV2_TYPE_CC110X,
+};
+
+/**
+ * @brief   Possible event types that are send from the device driver to the
+ *          upper layer
+ */
+typedef enum {
+    NETDEV2_EVENT_ISR,           /**< driver needs it's ISR handled */
+    NETDEV2_EVENT_RX_STARTED,    /**< started to receive a packet */
+    NETDEV2_EVENT_RX_COMPLETE,   /**< finished receiving a packet */
+    NETDEV2_EVENT_TX_STARTED,    /**< started to transfer a packet */
+    NETDEV2_EVENT_TX_COMPLETE,   /**< finished transferring packet */
+    /* expand this list if needed */
+} netdev2_event_t;
+
+typedef struct netdev2 netdev2_t;
+
+/**
+ * @brief   Event callback for signaling event to upper layers
+ *
+ * @param[in] type          type of the event
+ * @param[in] arg           event argument
+ */
+typedef void (*netdev2_event_cb_t)(netdev2_t *dev, netdev2_event_t event, void* arg);
+
+/**
+ * @brief Structure to hold driver state
+ *
+ * Supposed to be extended by driver implementations.
+ * The extended structure should contain all variable driver state.
+ */
+struct netdev2 {
+    const struct netdev2_driver *driver;    /**< ptr to that driver's interface. */
+    netdev2_event_cb_t event_callback;      /**< callback for device events */
+    void* isr_arg;                          /**< argument to pass on isr event */
+};
+
+/**
+ * @brief Structure to hold driver interface -> function mapping
+ *
+ * The send/receive functions expect/return a full ethernet
+ * frame (dst mac, src mac, ethertype, payload, no checksum).
+ */
+typedef struct netdev2_driver {
+    /**
+     * @brief Send frame
+     *
+     * @param[in] dev       network device descriptor
+     * @param[in] vector    io vector array to send
+     * @param[in] count     nr of entries in vector
+     *
+     * @return nr of bytes sent, or <=0 on error
+     */
+    int (*send)(netdev2_t *dev, const struct iovec *vector, int count);
+
+    /**
+     * @brief Get a received frame
+     *
+     * Supposed to be called from netdev2_event_handler().
+     *
+     * @param[in]   dev     network device descriptor
+     * @param[out]  buf     buffer to write into or NULL
+     * @param[in]   len     maximum nr. of bytes to read
+     *
+     * @return <=0 on error
+     * @return nr of bytes read if buf != NULL
+     * @return packet size if buf == NULL
+     */
+    int (*recv)(netdev2_t *dev, char* buf, int len);
```

Comment:
(as an out parameter) That would make the receive function you are having such problems with in #3683 much easier, because you don't need to copy twice.